Docs/api-gw-jwts-openshift-1-17-x (#19035)
hashicorp/consul: website/content/docs/connect/config-entries/api-gateway.mdx

- Default: `"default"` in Enterprise
- Data type: string

### `Listeners[].default`

> Specifies a block of default configurations to apply to the gateway listener. All routes attached to the listener inherit the default configurations. You can specify override configurations that have precedence over default configurations. Refer to the [`override` block](#listeners-override) for details.  

Reviewer: You can specify overrides on the gateway or on the HTTPRoute. Not sure if this is a good place to describe that, but thought I'd mention it

> Specifies a block of default configurations to apply to the gateway listener. All routes attached to the listener inherit the default configurations. You can specify override configurations that have precedence over default configurations. Refer to the [`override` block](#listeners-override) for details.  

Author: ```suggestion
Specifies a block of default configurations to apply to the gateway listener. All routes attached to the listener inherit the default configurations. You can specify override configurations that have precedence over default configurations in the [`override` block](#listeners-override) as well as in the `JWT` block in the [HTTP route configuration entry](/consul/docs/connect/config-entries/http-route).  
```

#### Values

- Default: None
- Data type: Map

### `Listeners[].default{}.JWT`

Specifies a block of default JWT verification configurations to apply to the gateway listener. You can specify override configurations that have precedence over default configurations. Refer to the [`override.JWT` block](#listeners-override) for details.

#### Values

- Default: None
- Data type: Map

### `Listeners[].default{}.JWT{}.Providers`

Specifies a list of default JWT provider configurations to apply to the gateway listener. A provider configuration contains the name of the provider and claims. You can specify override configurations that have precedence over default configurations. Refer to the [`override.JWT.Providers` block](#listeners-override-providers) for details.

#### Values

- Default: None
- Data type: List of maps

The following table describes the parameters you can specify in a member of the `Providers` list:

| Parameter | Description | Data type | Default |
| ---       | ---         | ---       | ---     |
| `Name` | Specifies the name of the provider. | String | None |

> | `VerifyClaim` | Specifies a list of paths and a value that define the claim. Consul verifies requests that match the claims declared in the listener JWT configuration and allow the request through the gateway. The `VerifyClaims` map specifies the following settings: <li>`Path`: List of paths to the claims</li><li>`Value`: ???</li> | Map | None |

Reviewer: ```suggestion
| `VerifyClaims` | Specifies a list of paths and a value that define the claim. Consul verifies requests that match the claims declared in the listener JWT configuration and allow the request through the gateway. The `VerifyClaims` map specifies the following settings: <li>`Path`: List of paths to the claims</li><li>`Value`: ???</li> | Map | None |
```

> | `VerifyClaim` | Specifies a list of paths and a value that define the claim. Consul verifies requests that match the claims declared in the listener JWT configuration and allow the request through the gateway. The `VerifyClaims` map specifies the following settings: <li>`Path`: List of paths to the claims</li><li>`Value`: ???</li> | Map | None |

Author: Yes, this is one of the areas I said I needed help with in the PR description. @nathancoleman @t-eckert , please add a suggestion to help fill in this gap. As far as the formatting, my thinking was that this block represented a tight enough unit of configuration to group into a table.

Refer to [Configure JWT verification settings](#configure-jwt-verification-settings) for an example configuration. 

### `Listeners[].override`

> Specifies a block of configurations to apply to the gateway listener. The override settings have  precedence over the [`default` configurations](#listeners-default).

Reviewer: Related to my above comment, the overrides here ultimately have precedence over the defaults defined on the `Gateway` resource and any JWT filter on the `HTTPRoute` resource.

Reviewer: I see the block below specifically mentions the route-level configuration, so maybe it doesn't make sense to mention those here as well. Just thinking out loud 😄

#### Values

- Default: None
- Data type: Map

### `Listeners[].override{}.JWT`

Specifies a block of JWT verification configurations to apply to the gateway listener. The override settings have  precedence over the [`default` configurations](#listeners-default), as well as any route-specific JWT configurations.

#### Values

- Default: None
- Data type: Map

### `Listeners[].override{}.JWT{}.Providers`

Specifies a list of JWT provider configurations to apply to the gateway listener. A provider configuration contains the name of the provider and claims. The override settings have precedence over defaults as well as any listener specific configuration. 

#### Values

- Default: None
- Data type: List of maps

The following table describes the parameters you can specify in a member of the `Providers` list:

| Parameter | Description | Data type | Default |
| ---       | ---         | ---       | ---     |
| `Name` | Specifies the name of the provider. | String | None |

> | `VerifyClaim` | Specifies a list of paths and a value that define the claim. Consul verifies requests that match the claims declared in the listener JWT configuration and allow the request through the gateway. The `VerifyClaims` map specifies the following settings: <ul><li>`Path`: List of paths to the claims</li><li>`Value`: ???</li></ul> | Map | None |

Reviewer: ```suggestion
| `VerifyClaims` | Specifies a list of paths and a value that define the claim. Consul verifies requests that match the claims declared in the listener JWT configuration and allow the request through the gateway. The `VerifyClaims` map specifies the following settings: <ul><li>`Path`: List of paths to the claims</li><li>`Value`: ???</li></ul> | Map | None |
```

Refer to [Configure JWT verification settings](#configure-jwt-verification-settings) for an example configuration. 

## Examples

Add the following example to the Examples section.

### Configure JWT verification settings

> The following example configures `listener-one` to verify that requests are from clients that have Okta user permissions by default. The listener also verifies claims that  <TODO - need help describing the example>:

Reviewer: ```suggestion
The following example configures `listener-one` to verify that requests include a token with Okta user permissions by default. The listener also verifies that the token has an audience of `api.apps.organization.com`.
```

It would probably also be helpful to describe that the user role verification can be overridden at the route level because it is only a `default`. The route could itself include a `JWTAuth` filter that verifies an Okta `admin` role instead.

In contrast, the audience claim verification will happen for every single request that comes through the listener, and it cannot be overridden at the route level because it is an `override` at the gateway level.

The order of precedence being
1. Gateway `override`
2. HTTPRoute filter
3. Gateway `default`
